README.md
# pyJARM

[![Latest version released on PyPi](https://img.shields.io/pypi/v/pyjarm.svg)](https://pypi.python.org/pypi/pyjarm)
[![License](https://img.shields.io/pypi/l/pyjarm)](https://github.com/PaloAltoNetworks/pyjarm/blob/main/LICENSE)
[![Black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)

## Overview

![pyjarm-gh](https://user-images.githubusercontent.com/7642165/105513920-33b09f00-5cd3-11eb-8dc7-e0b3cc9bd569.png)

pyJARM is a convenience library for the JARM fingerprinting tool. This library is based on the original python implementation [here](https://github.com/salesforce/jarm).

It requires python 3.7+.

## Installation
```
pip install pyjarm
```

## Usage

### Command Line
```
usage: jarm [-h] [-i INPUT] [-d] [-o OUTPUT] [-4] [-6] [-c [CONCURRENCY]]
            [--proxy PROXY] [--proxy-auth PROXY_AUTH] [--proxy-insecure]
            [--timeout TIMEOUT] [--suppress]
            [scan]

Enter an IP address/domain and port to scan or supply an input file.

positional arguments:
  scan                  Enter an IP or domain to scan.

optional arguments:
  -h, --help            show this help message and exit
  -i INPUT, --input INPUT
                        Provide a list of IP addresses or domains to scan, one
                        domain or IP address per line. Ports can be specified
                        with a colon (ex. 8.8.8.8:8443)
  -d, --debug           [OPTIONAL] Debug mode: Displays additional debug
                        details
  -o OUTPUT, --output OUTPUT
                        [OPTIONAL] Provide a filename to output/append results
                        to a CSV file.
  -4, --ipv4only        [OPTIONAL] Use only IPv4 connections (incompatible
                        with --ipv6only).
  -6, --ipv6only        [OPTIONAL] Use only IPv6 connections (incompatible
                        with --ipv4only).
  -c [CONCURRENCY], --concurrency [CONCURRENCY]
                        [OPTIONAL] Number of concurrent connections (default
                        is 2).
  --proxy PROXY         [OPTIONAL] Use proxy (format
                        http[s]://user:pass@proxy:port). HTTPS_PROXY env
                        variable is used by default if this is not set. Set
                        this to 'ignore' to ignore HTTPS_PROXY and use no
                        proxy.
  --proxy-auth PROXY_AUTH
                        [OPTIONAL] Send this header in Proxy-Authorization
                        (when using proxy).
  --proxy-insecure      [OPTIONAL] Do not verify SSL_CERTIFICATES (only when
                        HTTPS proxy is set).
  --timeout TIMEOUT     [OPTIONAL] Timeout to wait for connection attempts.
                        Default is 20 seconds
  --suppress            [OPTIONAL] Suppresses any exception or warning logging.
```

**Example**
```
$ pyjarm google.com
Target: google.com:443
JARM: 27d40d40d29d40d1dc42d43d00041d4689ee210389f4f6b4b5b1b93f92252d
```

### Scripted without asyncio
```
from jarm.scanner.scanner import Scanner

print(Scanner.scan("google.com", 443))
('27d40d40d29d40d1dc42d43d00041d4689ee210389f4f6b4b5b1b93f92252d', 'google.com', 443)
```

### Scripted with asyncio
```
import asyncio
from jarm.scanner.scanner import Scanner

print(asyncio.run(Scanner.scan_async("google.com", 443)))
('27d40d40d29d40d1dc42d43d00041d4689ee210389f4f6b4b5b1b93f92252d', 'google.com', 443)
```


## Contributors

- Andrew Scott - [andrew-paloalto](https://github.com/andrew-paloalto)
- Francesco Vigo - [fvigo](https://github.com/fvigo)
- Charlie Sestito - [csestito](http://github.com/csestito)


CHANGELOG.md
# Changelog

All notable changes to this this project will be documented in this file.

The format is based on [changelog.md](https://changelog.md/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.0.5] - 2021-02-08
- Add timeout and swallow error command line options
- Ensure that unsuccessful scan still return a valid JARM fingerprint (00000000000000000000000000000000000000000000000000000000000000)

## [0.0.4] - 2021-01-28
- Add improved command line support

## [0.0.3] - 2021-01-22
- minor fixes and updated documentation

## [0.0.2] - 2021-01-22
- Added proxy and async support

## [0.0.1] - 2021-01-15
- Initial release

jarm/cli.py
import argparse
import asyncio
from contextlib import suppress
from datetime import datetime, timezone
import logging

try:
    from jarm.constants import DEFAULT_TIMEOUT
    from jarm.scanner.scanner import Scanner
    from jarm.connection.connection import Connection
except ImportError:
    import os
    import sys

    sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
    from jarm.constants import DEFAULT_TIMEOUT
    from jarm.scanner.scanner import Scanner
    from jarm.connection.connection import Connection


def _scan(
    target: str,
    address_family: int = 0,
    proxy: str = None,
    proxy_auth: str = None,
    proxy_insecure: bool = None,
    concurrency: int = 2,
    timeout: int = DEFAULT_TIMEOUT,
    suppress: bool = False,
):
    if ":" in target:
        parts = target.split(":")
        host = parts[0]
        port = int(parts[1])
    else:
        host = target
        port = 443
    print(f"Target: {host}:{port}")
    results = asyncio.run(
        Scanner.scan_async(
            dest_host=host,
            dest_port=port,
            timeout=timeout,
            address_family=address_family,
            proxy=proxy,
            proxy_auth=proxy_auth,
            proxy_insecure=proxy_insecure,
            concurrency=concurrency,
            suppress=suppress,
        )
    )
    print(f"JARM: {results[0]}")
    return results


def run():
    parser = argparse.ArgumentParser(
        description="Enter an IP address/domain and port to scan or supply an input file."
    )
    group = parser.add_mutually_exclusive_group()
    group.add_argument("scan", nargs="?", help="Enter an IP or domain to scan.")
    group.add_argument(
        "-i",
        "--input",
        help="Provide a list of IP addresses or domains to scan, one domain or IP address per line. Ports can be specified with a colon (ex. 8.8.8.8:8443)",
        type=str,
    )
    parser.add_argument(
        "-d",
        "--debug",
        help="[OPTIONAL] Debug mode: Displays additional debug details",
        action="store_true",
    )
    parser.add_argument(
        "-o",
        "--output",
        help="[OPTIONAL] Provide a filename to output/append results to a CSV file.",
        type=str,
    )
    parser.add_argument(
        "-4",
        "--ipv4only",
        help="[OPTIONAL] Use only IPv4 connections (incompatible with --ipv6only).",
        action="store_true",
    )
    parser.add_argument(
        "-6",
        "--ipv6only",
        help="[OPTIONAL] Use only IPv6 connections (incompatible with --ipv4only).",
        action="store_true",
    )
    parser.add_argument(
        "-c",
        "--concurrency",
        nargs="?",
        help="[OPTIONAL] Number of concurrent connections (default is 2).",
        type=int,
    )
    parser.add_argument(
        "--proxy",
        help="[OPTIONAL] Use proxy (format http[s]://user:pass@proxy:port). HTTPS_PROXY env variable is used by default if this is not set. Set this to 'ignore' to ignore HTTPS_PROXY and use no proxy.",
        type=str,
    )
    parser.add_argument(
        "--proxy-auth",
        help="[OPTIONAL] Send this header in Proxy-Authorization (when using proxy).",
        type=str,
    )
    parser.add_argument(
        "--proxy-insecure",
        help="[OPTIONAL] Do not verify SSL_CERTIFICATES (only when HTTPS proxy is set).",
        action="store_true",
    )
    parser.add_argument(
        "--timeout",
        help="[OPTIONAL] Timeout to wait for connection attempts. Default is 20 seconds",
        type=int,
    )
    parser.add_argument(
        "--suppress",
        help="[OPTIONAL] Suppresses any exception logging.",
        action="store_true",
    )
    args = parser.parse_args()
    concurrency = args.concurrency if args.concurrency else 2
    if args.debug:
        logging.basicConfig(level=logging.DEBUG)
    if args.ipv4only and args.ipv6only:
        parser.error("Cannot specify both --ipv4only and --ipv6only at the same time")
    address_family = Connection.AddressFamily.AF_ANY  # either IPv4 or IPv6 allowed
    if args.ipv4only:
        address_family = Connection.AddressFamily.AF_INET
    elif args.ipv6only:
        address_family = Connection.AddressFamily.AF_INET6
    if args.scan is None and args.input is None:
        parser.error("A domain/IP to scan or an input file is required to run")
    elif args.scan is not None:
        results = [
            _scan(
                args.scan,
                address_family=address_family,
                proxy=args.proxy,
                proxy_auth=args.proxy_auth,
                proxy_insecure=args.proxy_insecure,
                concurrency=concurrency,
                timeout=args.timeout,
                suppress=args.suppress,
            )
        ]
    else:
        targets = []
        results = []
        with open(args.input, "r") as inpt:
            targets = [*inpt.read().splitlines()]
        for target in targets:
            results.append(
                _scan(
                    target,
                    address_family=address_family,
                    proxy=args.proxy,
                    proxy_auth=args.proxy_auth,
                    proxy_insecure=args.proxy_insecure,
                    concurrency=concurrency,
                    timeout=args.timeout,
                    suppress=args.suppress,
                )
            )
    if args.output is not None:
        with open(args.output, "w") as out:
            utc_now = datetime.now(timezone.utc).isoformat()
            out.write("Host,Port,JARM,ScanTime\n")
            for res in results:
                out.write(f"{res[1]},{res[2]},{res[0]},{utc_now}\n")
